xtc/src/ledger.rs

```diff
@@ -140,63 +193,127 @@ impl Ledger {
     }
 }
 
-#[derive(CandidType, Clone, Debug, PartialEq)]
-pub enum APIError {
+//////////////////// BEGIN OF ERC-20 ///////////////////////
+
+#[query(name=balanceOf)]
+pub async fn balance_of(account: Principal) -> Nat {
+    let ledger = ic_kit::get_context().get::<Ledger>();
+    Nat::from(ledger.balance(&account))
+}
+
+#[derive(Deserialize, CandidType)]
+pub struct TransferArguments {
+    pub to: Principal,
+    pub amount: u64,
+}
+
+#[derive(CandidType, Debug)]
+pub enum TransferError {
     InsufficientBalance,
-    InsufficientAllowance,
+    AmountTooLarge,
+    CallFailed,
     Unknown,
 }
 
-#[derive(CandidType, Clone, Debug)]
-pub struct ErrorDetails {
-    msg: &'static str,
-    code: APIError,
+#[query]
+pub async fn allowance(from: Principal, to: Principal) -> Nat {
+    return get_context().get::<Ledger>().allowance(&from, &to).into();
 }
 
-static InsufficientBalanceError: ErrorDetails = ErrorDetails {
-    msg: "Insufficient Balance",
-    code: APIError::InsufficientBalance,
-};
+#[update]
+pub async fn approve(to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+    use ic_cdk::export::candid;
+    let ic = get_context();
+    let caller = ic.caller();
+
+    crate::progress().await;
 
-static InsufficientAllowanceError: ErrorDetails = ErrorDetails {
-    msg: "Insufficient Allowance",
-    code: APIError::InsufficientAllowance,
-};
+    let ledger = ic.get_mut::<Ledger>();
+    let amount_u64: u64 =
+        utils::convert_nat_to_u64(amount).expect("Amount cannot be represented as u64");
+    let fee = compute_fee(amount_u64);
 
-static UnknownError: ErrorDetails = ErrorDetails {
-    msg: "Unknown",
-    code: APIError::Unknown,
-};
+    ledger.approve(&caller, &to, amount_u64, fee)?;
 
-#[query]
-fn totalSupply() -> Nat {
-    StatsData::get().supply
+    let transaction = Transaction {
+        timestamp: ic.time(),
+        cycles: amount_u64,
+        fee,
+        kind: TransactionKind::Approve {
+            from: caller,
+            to: to,
+        },
+    };
+
+    Ok(Nat::from(
+        ic.get_mut::<HistoryBuffer>().push(transaction.clone()),
+    ))
 }
 
-#[update]
-pub async fn balance(account: Option<Principal>) -> u64 {
+#[update(name=transferErc20)]
+pub async fn transfer_erc20(to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+
     let ic = get_context();
```

Comment:
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*